x/subscription/keeper/cu_tracker.go

```diff
@@ -5,34 +5,179 @@ import (
+		if !found {
+			utils.LavaFormatWarning("cannot remove cu tracker", legacyerrors.ErrKeyNotFound,
+				utils.Attribute{Key: "sub", Value: sub},
+				utils.Attribute{Key: "provider", Value: provider},
+				utils.Attribute{Key: "block", Value: strconv.FormatUint(entryBlock, 10)},
+			)
+		}
+		trackedCuList = append(trackedCuList, trackedCuInfo{
+			provider:  provider,
+			trackedCu: cu,
+			chainID:   chainID,
+			block:     entryBlock,
+		})
+		totalCuUsedBySub += cu
+	}
+
+	return trackedCuList, totalCuUsedBySub
+}
+
+// remove only before the sub is deleted
+func (k Keeper) RewardAndResetCuTracker(ctx sdk.Context, cuTrackerTimerKeyBytes []byte, cuTrackerTimerData []byte) {
+	sub := string(cuTrackerTimerKeyBytes)
+	shouldRemove := len(cuTrackerTimerData) != 0
+	trackedCuList, totalCuUsedBySub := k.getSubTrackedCuInfo(ctx, sub)
+
+	var block uint64
+	if len(trackedCuList) == 0 {
+		utils.LavaFormatWarning("no tracked CU", types.ErrCuTrackerPayoutFailed,
+			utils.Attribute{Key: "sub_consumer", Value: sub},
+		)
+		return
+	}
+
+	block = trackedCuList[0].block
+	plan, err := k.GetPlanFromSubscription(ctx, sub, block)
+	if err != nil {
+		utils.LavaFormatError("cannot find subscription's plan", types.ErrCuTrackerPayoutFailed,
+			utils.Attribute{Key: "sub_consumer", Value: sub},
+		)
+		return
+	}
+
+	for _, trackedCuInfo := range trackedCuList {
+		trackedCu := trackedCuInfo.trackedCu
+		provider := trackedCuInfo.provider
+		chainID := trackedCuInfo.chainID
+
+		err := k.resetCuTracker(ctx, sub, trackedCuInfo, shouldRemove)
+		if err != nil {
+			utils.LavaFormatError("removing/reseting tracked CU entry failed", err,
+				utils.Attribute{Key: "provider", Value: provider},
+				utils.Attribute{Key: "tracked_cu", Value: trackedCu},
+				utils.Attribute{Key: "chain_id", Value: chainID},
+				utils.Attribute{Key: "sub", Value: sub},
+				utils.Attribute{Key: "block", Value: ctx.BlockHeight()},
+			)
+			return
+		}
+
+		// monthly reward = (tracked_CU / total_CU_used_in_sub_this_month) * plan_price
+		// TODO: deal with the reward's remainder (uint division...)
+		totalMonthlyReward := plan.Price.Amount.MulRaw(int64(trackedCu)).QuoRaw(int64(totalCuUsedBySub))
+
+		// calculate the provider reward (smaller than totalMonthlyReward because it's shared with delegators)
+		providerAddr, err := sdk.AccAddressFromBech32(provider)
+		if err != nil {
+			utils.LavaFormatError("invalid provider address", err,
+				utils.Attribute{Key: "provider", Value: provider},
+			)
+			return
+		}
+
+		// Note: if the reward function doesn't reward the provider because he was unstaked, we only print an error and not returning
+		providerReward, err := k.dualstakingKeeper.RewardProvidersAndDelegators(ctx, providerAddr, chainID, totalMonthlyReward, types.ModuleName)
+		if err == dualstakingtypes.ErrProviderNotStaked {
+			utils.LavaFormatWarning("sending provider reward with delegations failed", err,
```

Comment:
Following our offline discussion, I changed the unresponsiveness punishment ("jail") to be freezing the provider instead of unstaking it. see https://github.com/lavanet/lava/commit/0b35ce0a7a270162105f7bf918cceb06f74d63dd